skimage/registration/_tvl1.py

```diff
@@ -0,0 +1,206 @@
+# coding: utf-8
+"""TV-L1 optical flow algorithm implementation.
+
+"""
+
+from functools import partial
+import numpy as np
+from scipy import ndimage as ndi
+from skimage.transform import warp
+
+from .utils import coarse_to_fine
+
+
+def _tvl1(I0, I1, flow0, dt, lambda_, tau, nwarp, niter, tol, prefilter):
+    """TV-L1 solver for optical flow estimation.
+
+    Parameters
+    ----------
+    I0 : ~numpy.ndarray
+        The first gray scale image of the sequence.
+    I1 : ~numpy.ndarray
+        The second gray scale image of the sequence.
+    flow0 : ~numpy.ndarray
+        Vector field initialization.
+    dt : float
+        Time step of the numerical scheme. Convergence is proved for
+        values dt < 0.125, but it can be larger for faster
+        convergence.
+    lambda_ : float
+        Attachement parameter. The smaller this parameter is,
+        the smoother is the solutions.
+    tau : float
+        Tightness parameter. It should have a small value in order to
+        maintain attachement and regularization parts in
+        correspondence.
+    nwarp : int
+        Number of times I1 is warped.
+    niter : int
+        Number of fixed point iteration.
+    tol : float
+        Tolerance used as stopping criterion based on the L² distance
+        between two consecutive values of (u, v).
+    prefilter : bool
+        whether to prefilter the estimated optical flow before each
+        image warp.
+
+    Returns
+    -------
+    flow : ~numpy.ndarray
+        The estimated optical flow.
+
+    """
+
+    grid = np.meshgrid(*[np.arange(n) for n in I0.shape], indexing='ij')
+
+    f0 = lambda_ * tau
+    f1 = dt / tau
+    tol *= I0.size
+
+    flow = flow0
+
+    g = np.zeros((I0.ndim, ) + I0.shape)
+    proj = np.zeros((I0.ndim, I0.ndim, ) + I0.shape)
+
+    s_g = [slice(None), ] * g.ndim
+    s_p = [slice(None), ] * proj.ndim
+    s_d = [slice(None), ] * (proj.ndim-2)
+
+    for _ in range(nwarp):
+        if prefilter:
+            flow = ndi.filters.median_filter(flow, [1]+I0.ndim*[3])
+
+        wI1 = warp(I1, grid+flow, mode='nearest')
+        grad = np.array(np.gradient(wI1))
+        NI = (grad*grad).sum(0)
+        NI[NI == 0] = 1
+
+        rho_0 = wI1 - I0 - (grad*flow0).sum(0)
+
+        for _ in range(niter):
+
+            # Data term
+
+            rho = rho_0 + (grad*flow).sum(0)
+
+            idx = abs(rho) <= f0 * NI
+
+            flow_ = flow
+
+            flow_[:, idx] -= rho[idx]*grad[:, idx]/NI[idx]
+
+            idx = ~idx
+            srho = f0 * np.sign(rho[idx])
+            flow_[:, idx] -= srho*grad[:, idx]
+
+            # Regularization term
+            flow = flow_.copy()
+
+            for idx in range(flow.shape[0]):
+                s_p[0] = idx
+                for _ in range(2):
```

Comment:
Why is this block run twice exactly?

Reply:
You are right. This is much like a parameter. In the original paper [1] it is set to 1 and in the improved version [2] it is set to 5. In this revision I set it as a variable in the code `reg_niter`, do you think that I should expose it as a parameter of the function?